pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,265 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package nodelabeller
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+
+	kubevirtv1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+	util "kubevirt.io/kubevirt/pkg/virt-handler/node-labeller/util"
+)
+
+const (
+	kvmPath = "/dev/kvm"
+)
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	kvmController     *device_manager.DeviceController
+	clientset         kubecli.KubevirtClient
+	host              string
+	namespace         string
+	logger            *log.FilteredLogger
+	clusterConfig     *virtconfig.ClusterConfig
+	hypervFeatures    supportedFeatures
+	hostCapabilities  supportedFeatures
+	queue             workqueue.RateLimitingInterface
+	supportedFeatures []string
+	cpuInfo           cpuInfo
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, clusterConfig *virtconfig.ClusterConfig, clientset kubecli.KubevirtClient, host, namespace string) (*NodeLabeller, error) {
+	n := &NodeLabeller{
+		kvmController: kvmController,
+		clientset:     clientset,
+		host:          host,
+		namespace:     namespace,
+		logger:        log.DefaultLogger(),
+		clusterConfig: clusterConfig,
+		queue:         workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+	err := n.loadCPUInfo()
+	if err != nil {
+		n.logger.Errorf("node-labeller could not load cpu info: " + err.Error())
+		return nil, err
+	}
+
+	err = n.loadHostSupportedFeatures()
+	if err != nil {
+		n.logger.Errorf("node-labeller could not load supported features: " + err.Error())
+		return nil, err
+	}
+
+	err = n.loadHostCapabilities()
+	if err != nil {
+		n.logger.Errorf("node-labeller could not load host capabilities: " + err.Error())
+		return nil, err
+	}
+
+	n.loadHypervFeatures()
+
+	return n, nil
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.queue.ShutDown()
+	n.logger.Infof("node-labeller is running")
+
+	n.clusterConfig.SetConfigModifiedCallback(func() {
+		n.queue.Add(n.host)
+	})
+
+	interval := 3 * time.Minute
+	go wait.JitterUntil(func() { n.queue.Add(n.host) }, interval, 1.2, true, stop)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stop)
+	}
+	//<-stop
```

Comment:
@rmohr, @davidvossel, @vladikr the tests are probably failing due to this line. If I comment is, then test passed locally. I am not sure why it is happening.